core/foundation/test/testRRangeCast.cxx

```diff
@@ -0,0 +1,37 @@
+#include "ROOT/RRangeCast.hxx"
+
+#include "gtest/gtest.h"
+
+class ClassA {
+public:
+   virtual ~ClassA() {}
+};
+class ClassB : public ClassA {
+};
+
+TEST(RRangeCast, CStyleArrayStatic)
+{
+
+   ClassB b;
+   const ClassA *arr[3] = {&b, &b, &b};
+   int n = 0;
+   for (auto *item : ROOT::RangeStaticCast<const ClassB *>(arr)) {
+      EXPECT_EQ(item, static_cast<const ClassB *>(arr[n]));
```

Comment:
Here, you are checking the pointers. However, the types (i.e. that the casting happens) aren't checked. See this:
```c++
root [0] class A{} a; class B : public A {} b;
root [3] &a == &b
(bool) false
root [4] static_cast<A*>(&b) == &b
(bool) true
```
That's why I suggested the `static_assert`.

Reply:
Ah okay, I thought this was implicitly done with the pointer comparison, but I checked now that it doesn't cover it because you can compare `A*` with `B*`. Now I brought in the static assert.